src/rgw/rgw_data_sync.cc

```diff
@@ -152,6 +152,44 @@ bool RGWReadDataSyncStatusMarkersCR::spawn_next()
   return true;
 }
 
+class RGWReadDataSyncRecoveringShardsCR : public RGWShardCollectCR {
+  static constexpr int MAX_CONCURRENT_SHARDS = 16;
+
+  RGWDataSyncEnv *env;
+
+  uint64_t max_entries;
+  set<int> shards;
+  set<int>::iterator iter;
+
+  string marker;
+  map<int, std::set<std::string>> &entries_map;
+
+ public:
+  RGWReadDataSyncRecoveringShardsCR(RGWDataSyncEnv *env, uint64_t _max_entries, set<int>& _shards,
+      map<int, std::set<std::string>>& _entries_map)
```

Comment:
since the only caller is always requesting all shards from `[0,num_shards)`, the use of `set<int> shards` seems like overkill - can you have it take `int num_shards` instead? see `RGWReadDataSyncStatusMarkersCR` as an example